lib/chef/provider/aws_network_interface.rb

```diff
@@ -0,0 +1,207 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+require 'cheffish'
+require 'date'
+require 'retryable'
+
+class Chef::Provider::AwsNetworkInterface < Chef::Provisioning::AWSDriver::AWSProvider
+  class NetworkInterfaceStatusTimeoutError < TimeoutError
+    def initialize(new_resource, initial_status, expected_status)
+      super("timed out waiting for #{new_resource} status to change from #{initial_status} to #{expected_status}!")
+    end
+  end
+
+  class NetworkInterfaceStatusTimeoutError < TimeoutError
+    def initialize(new_resource, initial_status, expected_status)
+      super("timed out waiting for #{new_resource} status to change from #{initial_status} to #{expected_status}!")
+    end
+  end
+
+  class NetworkInterfaceInvalidStatusError < RuntimeError
+    def initialize(new_resource, status)
+      super("#{new_resource} is in #{status} state!")
+    end
+  end
+
+  def action_create
+    eni = super
+
+    if !new_resource.machine.nil?
+      update_eni(eni)
+    end
+  end
+
+  protected
+
+  def create_aws_object
+    eni = nil
+    converge_by "create new #{new_resource} in #{region}" do
+      eni = new_resource.driver.ec2.network_interfaces.create(options)
+      eni.tags['Name'] = new_resource.name
+    end
+
+    converge_by "wait for new #{new_resource} in #{region} to become available" do
+      wait_for_eni_status(eni, :available)
+      eni
+    end
+  end
+
+  def update_aws_object(eni)
+    if options.has_key?(:subnet)
+      if Chef::Resource::AwsSubnet.get_aws_object(options[:subnet], resource: new_resource) != eni.subnet
+        raise "#{new_resource} subnet is #{new_resource.subnet}, but actual network interface has subnet set to #{eni.subnet_id}.  Cannot be modified!"
+      end
+    end
+
+    # TODO implement private ip reassignment
```

Comment:
I don't think this TODO is valid - I don't see in http://docs.aws.amazon.com/AWSRubySDK/latest/AWS/EC2/Client.html#modify_network_interface_attribute-instance_method how this could be updated